### ISSUE TITLE
upgrade all pipelines to latest version of Node / NPM / ES6 ("native")

### ISSUE THREAD

**Issue by thescientist13** [kenzanlabs/keystone#34]
Latest Node - v6.9
Latest NPM - v3.8

Rewrite all pipelines to use latest available ES6 features native in the latest release of Node.  This should include adding an engine key in _package.json_

cc: @cxoltero / @paul-barry-kenzan / @mgnsharon

**Comment by cxoltero** [comment]
When choosing minimum version of Node required, take into account that if we want to fully use ES6 syntax, latest release of Node 5.x only supports up to 59% of es6 syntax while ^6.9.2 supports 99%.
Node is also starting to promote users to transition to v4 or v6 ([Node 0.10.x 0.12.x support outlook](https://nodejs.org/en/blog/release/v6.0.0/#header-what-about-node-js-v0-10-and-v0-12)).

@obuckley-kenzan  | @bsawyer  | @paul-barry-kenzan

**Comment by thescientist13** [comment]
Yes, I would prefer v6 since it is the current LTS version.  I guess the main question is at this point to all is:

If we never set a minimum or expected version of Node either in our documentation or via _package.json_ does it constitute a breaking change?  What is everyone's thoughts on how the runtime should influence the code running on it, semver wise? 

@mgnsharon / @paul-barry-kenzan / @bsawyer

**Comment by bsawyer** [comment]
I would recommend checking min version support for features used: http://node.green/ and define those minimum requirements with `engine-strict` enabled. That way if someone tries to install under an unsupported runtime it will fail. For a versioning plan for updating the latest engine developed against and supported would it make sense to only define a max version if there were unsupported changes in a newer version of node? At which point I would think the question is whether or not failing on attempted installs was a bug-fix or functionality provided by the package in order to determine which level to bump the version, patch or minor respectively.

@thescientist13 @mgnsharon @cxoltero  @paul-barry-kenzan

**Comment by thescientist13** [comment]
Thanks @bsawyer 

I think I am ok making the call to go full ES6 and if we're making the effort to upgrade to a new NodeJS version as a result, it should be to the current LTS and set that as the minimum, to allow us to use any ES6 feature without friction or worry of having to bump the engine or the version number just to get a new feature and possibly force yet another version bump to the package.  

In thinking about it this way though per your comments, I could see this initial upgrade warranting a major version bump at the outset, to formalize the direction and reduce cognitive overhead around maintaining `engine` and for transparency to consumers.

That said... it would be great if there was a Node package that could statically analyze your code and tell you the right version of Node to set...

**Comment by thescientist13** [comment]
I think since Node v6 only started 10/2016, and given the nature of ES6 over ES5, I think at least for the jump in a pipeline, we should consider it a breaking change.  Features implemented after that bump can be taken on a case by case basis.
